File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.cpp
class Solution {
public:
    int minCost(int n, vector<int>& cuts) {
        int m = cuts.size();
        cuts.insert(cuts.begin(), 0);
        cuts.push_back(n);
        sort(cuts.begin(), cuts.end());
        
        vector<vector<int>> dp(m + 2, vector<int>(m + 2, 0));
        
        for (int diff = 2; diff <= m + 1; ++diff){
            for (int left = 0; left <= m + 1 - diff; ++left){
                int right = left + diff, ans = INT_MAX;
                
                for (int mid = left + 1; mid < right; ++mid)
                    ans = min(ans, dp[left][mid] + dp[mid][right] + cuts[right] - cuts[left]);
                
                dp[left][right] = ans;
            }
        }
        
        return dp[0][m + 1];
    }
};

File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.py
class Solution:
    def cost(self, left: int, right: int, memo: Dict[int, int], cuts: List[int]) -> int:
        if (left, right) in memo:
            return memo[(left, right)]
        
        if right - left == 1:
            return 0
        ans = min(self.cost(left, mid, memo, cuts) + self.cost(mid, right, memo, cuts) + cuts[right] - cuts[left] for mid in range(left + 1, right))
        memo[(left, right)] = ans
        return ans
    
    def minCost(self, n: int, cuts: List[int]) -> int:
        memo = dict()
        cuts = [0] + sorted(cuts) + [n]
        
        return self.cost(0, len(cuts) - 1, memo, cuts)